src/raft/recv_install_snapshot.c

```diff
@@ -328,34 +279,482 @@ static const struct sm_conf leader_states[LS_NR] = {
+}
+
+struct create_ht_data {
+	struct snapshot_leader_state *state;
+	int next_state;
+	struct raft_message *reply; /* owned */
+	struct raft_io_send *reply_req; /* owned */
+};
+
+static int create_ht_and_stmt(struct raft_io_async_work *req)
+{
+	char db_filename[30];
+	// TODO tracef with err_msg and tracef in general.
+	char *err_msg;
+	int rv;
+
+	struct create_ht_data *data = (struct create_ht_data *)req->data;
+	struct snapshot_leader_state *state = data->state;
+	sprintf(db_filename, "ht-%lld", state->follower_id);
+	rv = UvOsUnlink(db_filename);
+	assert(rv == 0 || errno == ENOENT);
+	rv = sqlite3_open_v2(db_filename, &state->ht, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, "unix");
+	assert(rv == SQLITE_OK);
+	rv = sqlite3_exec(state->ht,
+			  "CREATE TABLE map (checksum INTEGER NOT NULL, pageno "
+			  "INTEGER NOT NULL UNIQUE);",
+			  NULL, NULL, &err_msg);
+	assert(rv == SQLITE_OK);
+	rv = sqlite3_exec(state->ht, "CREATE INDEX map_idx on map(checksum);",
+			  NULL, NULL, &err_msg);
+	assert(rv == SQLITE_OK);
+	rv = sqlite3_prepare_v2(state->ht,
+				"INSERT OR IGNORE INTO map VALUES (?, ?);", -1,
+				&state->ht_stmt, NULL);
+	assert(rv == SQLITE_OK);
+
+	return SQLITE_OK;
+}
+
+static void create_ht_send_reply_cb(struct raft_io_send *req, int status) {
+	(void)status;
+
+	struct create_ht_data *data = (struct create_ht_data *)req->data;
+	sm_move(&data->state->sm, data->next_state);
+
+	raft_free(data->reply);
+	raft_free(data);
+	raft_free(req);
+}
+
+static void create_ht_send_reply(struct raft_io_async_work *req, int status) {
+	struct create_ht_data *data = (struct create_ht_data *)req->data;
+
+	if (status != 0) {
+		raft_free(data->reply);
+		raft_free(data);
+		raft_free(req);
+		return;
+	}
+
+	raft_free(req);
+	data->state->io->async_send_message(data->reply_req,
+			data->reply,
+			create_ht_send_reply_cb);
+}
+
+
+static void async_create_ht_and_reply(struct snapshot_leader_state *state, struct raft_message *reply, int next_state)
+{
+	struct create_ht_data *data;
+	data = raft_malloc(sizeof *data);
+	assert(data != NULL);
+	*data = (struct create_ht_data) {
+		.state = state,
+		.next_state = next_state,
+		.reply = reply,
+	};
+
+	struct raft_io_send *reply_req;
+	reply_req = raft_malloc(sizeof *reply_req);
+	assert(reply_req != NULL);
+	/* The data is reference circularly to be able to access it in the send
+	 * reply callback as well. */
+	reply_req->data = data;
+	data->reply_req = reply_req;
+
+	struct raft_io_async_work *work;
+	work = raft_malloc(sizeof *work);
+	assert(work != NULL);
+	*work = (struct raft_io_async_work) {
+		.work = create_ht_and_stmt,
+		.data = data,
+	};
+	state->io->async_work(work, create_ht_send_reply);
+}
+
+bool is_main_thread(void)
+{
+	// TODO: thread local storage.
 	return true;
 }
 
+void send_snapshot_cb(struct raft_io_send *req, int status) {
+	(void)req;
+	(void)status;
+	struct snapshot_leader_state *state = (struct snapshot_leader_state *) req->data;
```

Comment:
not sure if it is a good moment for this (but it's always not late to step on the path of samurai).

It looks more maintainable and better from perspective of code analysis (there's no guarantee that req->data lives the same live as the structure containing it) just to embed `struct raft_io_send` into `struct snapshot_leader_state` and use container_of() to obtain `struct snapshot_leader_state` back in the function.

Reply:
Wouldn't that be a problem in the context of scheduling async callbacks? We could have several `raft_io_send` when processing several messages. Or should we control that and not process more than 1 message?